beacon_node/network/src/network_beacon_processor/mod.rs

```diff
@@ -847,6 +849,73 @@ impl<T: BeaconChainTypes> NetworkBeaconProcessor<T> {
                 "error" => %e)
         });
     }
+
+    /// Attempt to reconstruct all data columns if the following conditions satisfies:
+    /// - Our custody requirement is all columns
+    /// - We have >= 50% of columns, but not all columns
+    ///
+    /// Returns `Some(AvailabilityProcessingStatus)` if reconstruction is successfully performed,
+    /// otherwise returns `None`.
+    async fn attempt_data_column_reconstruction(
+        &self,
+        block_root: Hash256,
+    ) -> Option<AvailabilityProcessingStatus> {
+        let result = self.chain.reconstruct_data_columns(block_root).await;
+        match result {
+            Ok(Some((availability_processing_status, data_columns_to_publish))) => {
+                self.send_network_message(NetworkMessage::Publish {
+                    messages: data_columns_to_publish
+                        .iter()
+                        .map(|d| {
+                            let subnet = DataColumnSubnetId::from_column_index::<T::EthSpec>(
+                                d.index as usize,
+                                &self.chain.spec,
+                            );
+                            PubsubMessage::DataColumnSidecar(Box::new((subnet, d.clone())))
+                        })
+                        .collect(),
+                });
+
+                match &availability_processing_status {
+                    AvailabilityProcessingStatus::Imported(hash) => {
+                        debug!(
+                            self.log,
+                            "Block components available via reconstruction";
+                            "result" => "imported block and custody columns",
+                            "block_hash" => %hash,
+                        );
+                        self.chain.recompute_head_at_current_slot().await;
+                    }
+                    AvailabilityProcessingStatus::MissingComponents(_, _) => {
+                        debug!(
+                            self.log,
+                            "Block components still missing block after reconstruction";
+                            "result" => "imported all custody columns",
+                            "block_hash" => %block_root,
+                        );
+                    }
+                }
+
+                Some(availability_processing_status)
+            }
+            Ok(None) => {
+                trace!(
+                    self.log,
+                    "Reconstruction not required for block";
+                    "block_hash" => %block_root,
+                );
+                None
```

Comment:
Consider we run this feature in production and reconstruction never happens. The next question would be why. The only evidence will be the absence of logs and all metrics being 0. We may want to log and track WHY we are not attempting reconstruction. How often is `attempt_data_column_reconstruction` called? We could add a metric  here with the reason why reconstruction is not done: i.e. not enough columns, not supernode etc. Could be noisy but I worry we are missing information if reconstruction just doesn't happen.

Reply:
Also we could have a debug log when importing a data column into the da_checker that prints the current count of imported valid data columns for that block root, and the expected custody requirement. That way we can track if we ever reached the reconstruction threshold or not.

Reply:
I've added metric `KZG_DATA_COLUMN_RECONSTRUCTION_INCOMPLETE_TOTAL` for this. I think logging would be very noisy, especially after the getBlobsV1 optimisation.

Reply:
Added reconstruction metric and da checker debug log in https://github.com/sigp/lighthouse/pull/6403/commits/0e6eaa227890194983cd523d5e6be498907dfdad